### PR TITLE
Fix up parsing of datetime string with microseconds.

### LINKED ISSUE #17601: DateTime microseconds make it null on save
### Description

I had an old app migrated to cake5 and was wondering why a test was failing for save()
Turns out the datetime string data posted including microseconds is not accepted

Debugging in beforeMarshal()
```
'start_time' => '2024-03-02 15:46:00.000000',
'end_time' => '2024-03-02 19:16:00',
```
The mysql table field is both "datetime", the 2nd gets saved, the first one becomes NULL

Which is suprising as cake's own https://github.com/cakephp/cakephp/blob/1daf3c0854c139bac85d3535505cef6154f9d164/src/Validation/Validation.php#L1704 has logic to parse arrays into string this way:
```php
                $formatted .= sprintf(
                    '%02d:%02d:%02d.%06d',
                    $value['hour'],
                    $value['minute'],
                    $value['second'],
                    $value['microsecond']
                );
```

It literally generates the

    2024-03-02 15:46:00.000000

### CakePHP Version

5.x

### PHP Version

8.2

### PR DESCRIPTION
**Pull request by dereuromark**
Resolves https://github.com/cakephp/cakephp/issues/17601

### CODE REVIEW

**Comment by othercorey** [comment]
Looks ok. I think this was skipped because DateTimeType fields do not use fractional seconds. Users have to map DateTimeFractionalType for that.

But, we can parse and ignore the milliseconds.

**Comment by dereuromark** [comment]
Exactly, I would expect it to be at least accepted for the rest of the input, instead of completely ignored.

**Comment by othercorey** [comment]
> Exactly, I would expect it to be at least accepted for the rest of the input, instead of completely ignored.

Is it ignored because i's not used when converting to database format.

**Comment by dereuromark** [comment]
I mean the rest of the input, so the string before the microseconds.

File: src/Database/Type/DateTimeType.php [hunk 1]
```diff
@@ -53,6 +53,8 @@ class DateTimeType extends BaseType implements BatchCastingInterface
         'Y-m-d\TH:i',
         'Y-m-d\TH:i:s',
         'Y-m-d\TH:i:sP',
+        'Y-m-d H:i:s.P',
+        'Y-m-d H:i:s.u',
```
Comment (ADmad) [inline src/Database/Type/DateTimeType.php:57]: A datetime column with microseconds should be mapped to the `DateTimeFractionalType` class, which already has this [format](https://github.com/cakephp/cakephp/blob/5.x/src/Database/Type/DateTimeFractionalType.php#L37) so it's not needed here.

Comment (othercorey) [reply src/Database/Type/DateTimeType.php:57]: That was the original assumption. However, DateTimeFractionalType will create a string with the fraction and that probably breaks the mysql parser.

Comment (ADmad) [reply src/Database/Type/DateTimeType.php:57]: If you think this is fine/necessary then I have no issues. I haven't used microseconds in practice so not familiar with it's intricacies.

Comment (othercorey) [reply src/Database/Type/DateTimeType.php:57]: It's definitely unexpected input, but what can you do.

In the future, I think we might simplify these types so they parse anything reasonable and only care about how they convert to db format and timezone.

Comment (dereuromark) [reply src/Database/Type/DateTimeType.php:57]: As explained in the ticket, this about the datetime value part of the value, not the milliseconds.
The expectation is that this first part is not dropped silently.
That the ms cannot be stored is fine here.

Comment (markstory) [reply src/Database/Type/DateTimeType.php:57]: I think it makes sense to parse the microseconds into the PHP object but not persist that precision.

**Pull request merged**